server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/BuildComponentTreeStep.java

```diff
@@ -64,64 +66,52 @@ public String getDescription() {
 
   @Override
   public void execute() {
-    String branch = analysisMetadataHolder.getBranch();
-    ScannerReport.Component reportProject = reportReader.readComponent(analysisMetadataHolder.getRootComponentRef());
-    String projectKey = createKey(reportProject.getKey(), branch);
-    UuidFactory uuidFactory = new UuidFactory(dbClient, projectKey);
-
     try (DbSession dbSession = dbClient.openSession(false)) {
-      BaseAnalysisSupplier baseAnalysisSupplier = new BaseAnalysisSupplier(dbClient, dbSession);
-      ComponentRootBuilder rootBuilder = new ComponentRootBuilder(branch,
-        uuidFactory::getOrCreateForKey,
+      ScannerReport.Component reportProject = reportReader.readComponent(analysisMetadataHolder.getRootComponentRef());
+      ComponentKeyGenerator keyGenerator = loadKeyGenerator();
+
+      // root key of branch, not necessarily of project
+      String rootKey = keyGenerator.generateKey(reportProject, reportProject);
+
+      // loads the UUIDs from database. If they don't exist, then generate new ones
+      ComponentUuidFactory componentUuidFactory = new ComponentUuidFactory(dbClient, dbSession, rootKey);
+
+      String rootUuid = componentUuidFactory.getOrCreateForKey(rootKey);
+      SnapshotDto baseAnalysis = loadBaseAnalysis(dbSession, rootUuid);
+
+      ComponentTreeBuilder builder = new ComponentTreeBuilder(keyGenerator,
+        componentUuidFactory::getOrCreateForKey,
         reportReader::readComponent,
-        () -> dbClient.componentDao().selectByKey(dbSession, projectKey),
-        baseAnalysisSupplier);
-      Component project = rootBuilder.build(reportProject, projectKey);
+        analysisMetadataHolder.getProject(),
+        baseAnalysis);
+      Component project = builder.buildProject(reportProject);
+
       treeRootHolder.setRoot(project);
-      analysisMetadataHolder.setBaseAnalysis(toAnalysis(baseAnalysisSupplier.apply(project.getUuid())));
+      analysisMetadataHolder.setBaseAnalysis(toAnalysis(baseAnalysis));
     }
   }
 
-  /**
-   * A supplier of the base analysis of the project (if it exists) that will cache the retrieved SnapshotDto and
-   * implement a sanity check to ensure it is always call with the same UUID value (since it's the project's UUID, it
-   * is unique for a whole task).
-   */
-  private static final class BaseAnalysisSupplier implements Function<String, Optional<SnapshotDto>> {
-    private final DbClient dbClient;
-    private final DbSession dbSession;
-    private String projectUuid = null;
-    private Optional<SnapshotDto> cache = null;
-
-    private BaseAnalysisSupplier(DbClient dbClient, DbSession dbSession) {
-      this.dbClient = dbClient;
-      this.dbSession = dbSession;
-    }
+  private ComponentKeyGenerator loadKeyGenerator() {
+    return Stream.of(analysisMetadataHolder.getBranch(), Optional.of(new DefaultKeyGenerator()))
+      // TODO pull request generator will be added here
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)